### ISSUE TITLE
Sporadic BackgroundScanHandler.waitForScanToComplete race / broken

### ISSUE THREAD

**Issue by Player3324** [MinecraftForge/MinecraftForge#8156]
**Minecraft Version:** 1.17.1

**Forge Version:** 37.0.90

**Logs:** https://gist.github.com/sfPlayer1/c0adf26ce31c5a5828b51d62be0f56f6

**Steps to Reproduce:**
 1. Add enough mods for the scan to take longer than 50 ms?
 2. Start the server often enough to catch this since it is timing sensitive

**Description of issue:**

The server (FC1) sporadically crashes with an IllegalStateException instead of loading mods as usual. From what I gathered this is caused by a bug in the implementation of `net.minecraftforge.fml.loading.moddiscovery.BackgroundScanHandler.waitForScanToComplete()`.

There are two closely related issues, assuming the code is intended to not have an overall time limit, but just tries to invoke the ticker regularly:
1. `modContentScanner.isShutdown()` is (presumably) always true due to `modContentScanner.shutdown()` above, it is likely meant to test `modContentScanner.isTerminated()` instead - otherwise the loop always terminates and serves no purpose
2. With 1. fixed there is a race between `awaitTermination` and the loop exit, `succeeded` won't be set appropriately if it terminates between the two tests

A potential solution could look like this (CC0):
```java
modContentScanner.shutdown();

try {
    do {
        ticker.run();
    } while (!modContentScanner.awaitTermination(50, TimeUnit.MILLISECONDS));
} catch (InterruptedException e) {
    Thread.currentThread().interrupt();
    if (!modContentScanner.isTerminated()) throw new IllegalStateException("Failed to complete mod scan"); // not sure this is ideal, but it replicates the old logic
}
```

**Comment by Tamaized** [comment]
For what it's worth, I get this over 50% of the time I try to launch the game in my dev workspace and that's just the one mod I'm working on being loaded, nothing extra.. To clarify, it's with any mod I'm working on, not one specific one.

Extremely annoying.

**Comment by ChampionAsh5357** [comment]
For those who are having this issue, we would like you to test #8176 to see if you can encounter the same situation again or if any new issues arise. Thank you.

**Comment by Drullkus** [comment]
Figured this info would be helpful - A fellow developer was hitting this problem multiple times before he'd get a successful load up. After conversing I discovered he had only a HDD which I recommended him to add a NVMe SSD to his system.

After he upgraded, his local repo was still located on his HDD but his system's pagefile was set for his SSD. There's probably some correlation there but I'm not knowledgeable enough as for specifics. So far of all of the times he's successfully launched, he's by far only had one Mod Scan failure.
